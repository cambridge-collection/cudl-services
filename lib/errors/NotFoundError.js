var BaseError = require('./BaseError');

module.exports = function NotFoundError(message, extra) {
    NotFoundError.super_.call(this, message, extra);
};

require('util').inherits(module.exports, BaseError);
