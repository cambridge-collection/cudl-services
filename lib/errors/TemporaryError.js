const BaseError = require('./BaseError');

class TemporaryError extends BaseError { }
module.exports = TemporaryError;
