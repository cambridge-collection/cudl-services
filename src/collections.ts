import {BasePostgresDAO} from './db';
import {Resource} from './resources';

export interface Collection {
  title: string;
  collectionID: string;
  collectionOrder: number;
}

export type GetItemCollections = (itemID: string) => Promise<Collection[]>;

export interface CollectionDAO extends Resource {
  getItemCollections: GetItemCollections;
}

export class PostgresCollectionDAO
  extends BasePostgresDAO
  implements CollectionDAO
{
  async getItemCollections(itemID: string): Promise<Collection[]> {
    const sql = `\
WITH RECURSIVE collection_membership(collectionid, title, collectionorder) AS (
  SELECT collections.collectionid, title, collectionorder, parentcollectionid
  FROM collections
         JOIN itemsincollection on collections.collectionid = itemsincollection.collectionid
  WHERE itemsincollection.itemid = $1::text AND itemsincollection.visible
  UNION
  SELECT collections.collectionid, collections.title, collections.collectionorder, collections.parentcollectionid
  FROM collections,
       collection_membership
  WHERE collections.collectionid = collection_membership.parentcollectionid
)
SELECT *
FROM collection_membership;`;
    return (await this.db.query(sql, [itemID])).rows.map(row => ({
      title: row.title,
      collectionID: row.collectionid,
      collectionOrder: Number(row.collectionorder),
    }));
  }
}
