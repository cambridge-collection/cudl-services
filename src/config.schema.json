{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "properties": {
    "darwinXTF": {"type": "string"},
    "dataDir": {"type": "string"},
    "legacyDcpDataDir": {"type": "string"},
    "postDatabase": {"type": "string"},
    "postHost": {"type": "string"},
    "postPort": {"type": "number"},
    "postPass": {"type": "string"},
    "postUser": {"type": "string"},
    "users": {
      "type": "object",
      "additionalProperties": {
        "type": "object",
        "properties": {
          "username": {"type": "string"},
          "email": {"type": "string"}
        },
        "required": ["username", "email"]
      }
    },
    "xtfBase": {"type": "string"},
    "xtfIndexPath": {"type": "string"}
  },
  "required": ["darwinXTF", "dataDir", "postDatabase", "postHost", "postPass", "postUser", "users", "xtfBase", "xtfIndexPath"]
}
