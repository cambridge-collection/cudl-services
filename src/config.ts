export const CONFIG_ENVAR = 'CUDL_SERVICES_CONFIG';

export function loadConfigFromEnvar(): Config {
  const configModulePath = process.env[CONFIG_ENVAR];
  if (!configModulePath) {
    throw new Error(`\
Configuration not found: envar ${CONFIG_ENVAR} be set to the path of the \
config module to load`);
  }

  return require(configModulePath);
}

export interface Config {
  dataDir: string;
  users: Users;
  darwinXTF: string;
  [index: string]: unknown;
}

export interface Users {
  [apiKey: string]: User;
}

export interface User {
  username: string;
  email: string;
}
