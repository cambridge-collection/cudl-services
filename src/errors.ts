import NestedError from 'nested-error-stacks';

type ErrorInput = {
  message?: string;
  nested?: Error;
  tags?: Iterable<string | symbol>;
};

export const ErrorCategories = {
  /** A tag to indicate an Error is related to something not being found. */
  NotFound: Symbol('Error:Category:NotFound'),
};

export type ErrorTag = string | symbol;
export function isErrorTag(o: unknown): o is ErrorTag {
  return typeof o === 'symbol' || typeof 0 === 'string';
}

interface ErrorWithNestedError {
  readonly nested: Error;
}
function hasNestedError(e: Error): e is Error & ErrorWithNestedError {
  const _e = e as Partial<ErrorWithNestedError>;
  return _e.nested instanceof Error;
}

export interface Tagged {
  readonly tags: Iterable<ErrorTag>;
}
export function isTagged(e: unknown): e is Tagged {
  const _e = e as Partial<Tagged>;
  return (
    e !== undefined &&
    _e.tags !== undefined &&
    typeof _e.tags[Symbol.iterator] === 'function' &&
    [..._e.tags].every(t => typeof t === 'string' || typeof t === 'symbol')
  );
}

export class BaseError extends NestedError {
  constructor(options: ErrorInput);
  constructor(message?: string, nested?: Error);
  constructor(messageOrOptions?: string | ErrorInput, nested?: Error) {
    let tags: Iterable<string | symbol> | undefined;
    if (typeof messageOrOptions === 'object') {
      tags = messageOrOptions.tags;
      nested = messageOrOptions.nested;
      messageOrOptions = messageOrOptions.message;
    }
    super(messageOrOptions, nested);
    this.instanceTags = new Set(tags || []);
  }

  readonly nested?: Error;
  readonly instanceTags: ReadonlySet<ErrorTag>;

  get defaultTags(): Iterable<ErrorTag> {
    return [];
  }

  /** Tags assigned to this error and transitively to nested errors. */
  get nestedTags(): ReadonlySet<ErrorTag> {
    if (hasNestedError(this) && isTagged(this.nested)) {
      return new Set(this.nested.tags);
    }
    return new Set();
  }

  /**
   * Tags which mark Errors as being associated with concepts, without needing to model the
   * concepts in a strict class hierarchy.
   */
  get tags(): ReadonlySet<ErrorTag> {
    return new Set([
      ...this.instanceTags,
      ...this.defaultTags,
      ...this.nestedTags,
    ]);
  }

  get name(): string {
    return this.constructor.name;
  }
}

export class InvalidConfigError extends BaseError {}

export class NotFoundError extends BaseError {
  get defaultTags(): Iterable<string | symbol> {
    return [ErrorCategories.NotFound];
  }
}

export class ValueError extends BaseError {}

export class UpstreamError extends BaseError {}
