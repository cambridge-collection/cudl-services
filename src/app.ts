import { XSLTExecutor } from '@lib.cam/xslt-nailgun';
import bodyParser from 'body-parser';
import Debugger from 'debug';
import express, { NextFunction, Request, Response } from 'express';
import passport from 'passport';
import path from 'path';
import { CollectionDAO, PostgresCollectionDAO } from './collections';
import { Config, User, Users } from './config';

import { DAOPool, PostgresDatabasePool } from './db';
import {
  CUDLMetadataRepository,
  DefaultCUDLMetadataRepository,
  LegacyDarwinMetadataRepository,
} from './metadata';
import { BaseResource, ExternalResources, using } from './resources';
import * as darwin from './routes/darwin';
import * as membership from './routes/membership';
import * as metadata from './routes/metadata';
import * as similarity from './routes/similarity';
import * as tags from './routes/tags';
import { PostgresTagsDAO, TagsDAO } from './routes/tags-impl';
import * as transcription from './routes/transcription';
import * as translation from './routes/translation';
import { DefaultXTF, XTF } from './xtf';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');
const Strategy = require('passport-accesstoken').Strategy;

const debug = Debugger('cudl-services');

// FIXME: move this to server?
//cache directories
// fs.ensureDirSync(config.cacheDir);
// fs.ensureDirSync(config.cacheDir+'/transcriptions');
// fs.ensureDirSync(config.cacheDir+'/translations');

export interface AppOptions {
  users: Users;
  metadataRepository: CUDLMetadataRepository;
  legacyDarwinMetadataRepository: LegacyDarwinMetadataRepository;
  collectionsDAOPool: DAOPool<CollectionDAO>;
  tagsDAOPool: DAOPool<TagsDAO>;
  darwinXtfUrl: string;
  xtf: XTF;
}

export class App extends BaseResource {
  readonly options: AppOptions;
  readonly expressApp: express.Application;
  protected readonly xsltExecutor: XSLTExecutor;

  constructor(options: AppOptions) {
    super();
    this.options = Object.freeze({ ...options });
    this.xsltExecutor = XSLTExecutor.getInstance();
    this.expressApp = this.createExpressApp();
  }

  static fromConfig(config: Config): ExternalResources<App> {
    const dbPool = PostgresDatabasePool.fromConfig(config);

    return new ExternalResources(
      new App({
        metadataRepository: new DefaultCUDLMetadataRepository(config.dataDir),
        legacyDarwinMetadataRepository: new LegacyDarwinMetadataRepository(
          config.legacyDcpDataDir
        ),
        collectionsDAOPool: PostgresCollectionDAO.createPool(dbPool),
        tagsDAOPool: PostgresTagsDAO.createPool(dbPool),
        users: config.users,
        darwinXtfUrl: config.darwinXTF,
        xtf: new DefaultXTF(config),
      }),
      [dbPool]
    );
  }

  private createExpressApp(): express.Application {
    const app = express();

    passport.use(
      new Strategy((token: string, done: (err: any, user: any) => void) => {
        process.nextTick(() => {
          const user = findByApiKey(this.options.users, token);
          return done(null, user || false);
        });
      })
    );

    // view engine setup
    app.set('views', path.resolve(__dirname, '../views'));
    app.set('view engine', 'pug');

    app.use(
      favicon(path.resolve(__dirname, '../public/images/brand/favicon.ico'))
    );
    app.use(logger('dev'));
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(cookieParser());
    app.use(passport.initialize());
    app.use(express.static(path.resolve(__dirname, '../public')));

    // Middleware to redirect trailing slashes to same URL without trailing slash
    app.use((req, res, next) => {
      if (req.url.substr(-1) === '/' && req.url.length > 1) {
        res.redirect(301, req.url.slice(0, -1));
      } else {
        next();
      }
    });

    app.use(
      '/v1/metadata',
      metadata.getRoutes({
        metadataRepository: this.options.metadataRepository,
      })
    );

    app.use(
      '/v1/rdb/membership',
      membership.getRoutes({
        getItemCollections: async (itemID: string) =>
          using(this.options.collectionsDAOPool.getInstance(), dao =>
            dao.getItemCollections(itemID)
          ),
      })
    );

    app.use('/v1/tags', tags.getRoutes({ daoPool: this.options.tagsDAOPool }));

    app.use(
      '/v1/transcription',
      transcription.getRoutes({
        metadataRepository: this.options.metadataRepository,
        legacyDarwinMetadataRepository: this.options
          .legacyDarwinMetadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/translation',
      translation.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/xtf/similarity',
      similarity.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xtf: this.options.xtf,
      })
    );

    app.use(
      '/v1/darwin',
      passport.authenticate('token', { session: false }),
      darwin.getRoutes({ darwinXtfUrl: this.options.darwinXtfUrl })
    );

    // 404 if no route matched
    app.use((req: Request, res: Response, next: NextFunction) => {
      res.status(404).send('Not Found');
    });

    return app;
  }

  async close(): Promise<void> {
    await Promise.all([super.close(), this.xsltExecutor.close()]);
  }
}

function findByApiKey(users: Users, apiKey: string): User | null {
  if (apiKey in users) {
    return users[apiKey];
  }
  return null;
}
