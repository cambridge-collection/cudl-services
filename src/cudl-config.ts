import Ajv, {ValidateFunction} from 'ajv';
import Debugger from 'debug';
import deepmerge from 'deepmerge';
import fs from 'fs';
import glob from 'glob';
import json5 from 'json5';
import util, {promisify} from 'util';
import {BaseError, InvalidConfigError} from './errors';

import fullConfigSchema from './full-config.schema.json';
import partialConfigSchema from './partial-config.schema.json';
import {NonOptional, requireNotUndefined} from './util';
import {Config} from './config';
import {App, Application, User, Users} from './app';
import {ExternalResources} from './resources';
import {MetadataProviderCUDLMetadataRepository} from './metadata/cudl';
import {FilesystemDataStore} from './metadata/filesystem';
import {PostgresCollectionDAO} from './collections';
import {PostgresTagsDAO} from './routes/tags-impl';
import {DefaultXTF, XTFConfig} from './xtf';
import {URL} from 'url';
import {DatabaseConfig, PostgresDatabasePool} from './db';

const debug = Debugger('cudl-services:config');

export const CONFIG_JSON_ENVAR = 'NODE_CONFIG';
export const CONFIG_FILE_ENVAR = 'NODE_CONFIG_FILE';
export const DEFAULT_CONFIG_GLOBS =
  '/etc/cudl-services/config.json?(5):/etc/cudl-services/conf.d/*.json?(5)';

const ajv = new Ajv({schemas: [fullConfigSchema, partialConfigSchema]});
const fullConfigValidator = requireNotUndefined(
  ajv.getSchema('full-config.schema.json')
);
const partialConfigValidator = requireNotUndefined(
  ajv.getSchema('partial-config.schema.json')
);

const DEFAULT_ZACYNTHIUS_SERVICE_URL =
  'http://codex-zacynthius-transcription.cudl.lib.cam.ac.uk';

export function splitEnvarPaths(paths: string | undefined) {
  return (paths || '').split(':').filter(p => !!p);
}

export async function loadConfigFile(
  filePath: string
): Promise<PartialCUDLConfigData> {
  let jsonDocument;
  try {
    jsonDocument = await promisify(fs.readFile)(filePath, 'utf-8');
  } catch (e) {
    throw new BaseError(
      `Failed to load config from file ${util.inspect(filePath)}: ${e}`,
      e
    );
  }

  return parseConfigFromJSON(jsonDocument, `file ${util.inspect(filePath)}`);
}

export function parseConfigFromJSON(
  jsonDocument: string,
  sourceDescription: string
): PartialCUDLConfigData {
  try {
    const config = json5.parse(jsonDocument);
    validateObjectIsPartialConfig(config);
    return config;
  } catch (e) {
    throw new BaseError(
      `Failed to load config from ${sourceDescription}: ${e}`,
      e
    );
  }
}

export interface ConfigSource {
  source: string;
  config: PartialCUDLConfigData;
}

export const DEFAULT_CONFIG: ConfigSource = Object.freeze({
  source: 'default values',
  config: Object.freeze({
    postPort: 5432,
    zacynthiusServiceURL: DEFAULT_ZACYNTHIUS_SERVICE_URL,
  }),
});

export function mergeConfigs(
  configSources: ConfigSource[]
): StrictCUDLConfigData {
  debug('Merging config from sources:', configSources);
  const merged = configSources
    .map(cs => cs.config)
    .reduce((a, b) => deepmerge(a, b));
  try {
    validateObjectIsFullConfig(merged);
    return merged;
  } catch (e) {
    const sourceDesc = configSources.map(cs => cs.source).join(', ');
    throw new InvalidConfigError(
      `Failed to load config: Result of merging ${sourceDesc} was not a valid config: ${e}`,
      e
    );
  }
}

export async function configPathsFromEnvar() {
  const envarGlobs = process.env[CONFIG_FILE_ENVAR];
  const defaultPatternUsed = envarGlobs === undefined;
  const globPatterns = splitEnvarPaths(
    defaultPatternUsed ? DEFAULT_CONFIG_GLOBS : envarGlobs
  );
  const paths = await Promise.all(globPatterns.map(p => promisify(glob)(p)));

  return {
    defaultPatternUsed,
    patterns: globPatterns,
    paths: paths.reduce((a, b) => a.concat(b)),
  };
}

export async function loadConfigFromEnvar(): Promise<StrictCUDLConfigData> {
  const configFilePathInfo = await configPathsFromEnvar();
  debug(
    `${
      configFilePathInfo.defaultPatternUsed
        ? `Envar ${CONFIG_FILE_ENVAR} is not set - using default config file locations:`
        : `Using config file locations from envar ${CONFIG_FILE_ENVAR}:`
    } ${configFilePathInfo.patterns.join(':')}; found ${
      configFilePathInfo.paths.length
    } files`,
    configFilePathInfo.paths
  );

  let configSources = [DEFAULT_CONFIG];
  configSources = configSources.concat(
    await Promise.all(
      configFilePathInfo.paths.map(async file => ({
        source: `file ${util.inspect(file)}`,
        config: await loadConfigFile(file),
      }))
    )
  );

  const configJSONEnvarDocument = process.env[CONFIG_JSON_ENVAR];
  if (configJSONEnvarDocument) {
    const source = `envar ${CONFIG_JSON_ENVAR}`;
    configSources.push({
      source,
      config: parseConfigFromJSON(configJSONEnvarDocument, source),
    });
  }

  if (configSources.filter(s => s !== DEFAULT_CONFIG).length === 0) {
    throw new Error('No configuration sources found');
  }

  return mergeConfigs(configSources);
}

function validateConfig(configValidator: ValidateFunction, config: unknown) {
  const valid = configValidator(config);
  if (!valid) {
    const message =
      configValidator?.errors?.map(err => err.message).join('; ') ||
      '* validation failed without messages *';
    throw new InvalidConfigError(`config is invalid: ${message}`);
  }
}

export function validateObjectIsPartialConfig(
  config: unknown
): asserts config is PartialCUDLConfigData {
  validateConfig(partialConfigValidator, config);
}

export function validateObjectIsFullConfig(
  config: unknown
): asserts config is StrictCUDLConfigData {
  validateConfig(fullConfigValidator, config);
}

export interface CUDLConfigData<U = Users> extends XTFConfig, DatabaseConfig {
  dataDir: string;
  users: U;
  darwinXTF: string;
  zacynthiusServiceURL: string;
}

export type PartialCUDLConfigData = Partial<
  CUDLConfigData<Users<Partial<User>>>
>;
export type StrictCUDLConfigData = NonOptional<CUDLConfigData>;

class ApplicationWithResources implements Application {
  protected readonly appWithResource: ExternalResources<Application>;

  constructor(appWithResource: ExternalResources<Application>) {
    this.appWithResource = appWithResource;
  }

  get expressApp() {
    return this.appWithResource.value.expressApp;
  }

  async close() {
    await this.appWithResource.close();
  }
}

class CUDLConfig implements Config {
  async createApplication(): Promise<Application> {
    return new ApplicationWithResources(
      this.createApplicationFromConfigData(await loadConfigFromEnvar())
    );
  }

  createApplicationFromConfigData(config: CUDLConfigData) {
    const dbPool = PostgresDatabasePool.fromConfig(config);

    return new ExternalResources(
      new App({
        metadataRepository: MetadataProviderCUDLMetadataRepository.forDataStore(
          new FilesystemDataStore(config.dataDir)
        ),
        collectionsDAOPool: PostgresCollectionDAO.createPool(dbPool),
        tagsDAOPool: PostgresTagsDAO.createPool(dbPool),
        users: config.users,
        darwinXtfUrl: config.darwinXTF,
        xtf: new DefaultXTF(config),
        zacynthiusServiceURL: new URL(config.zacynthiusServiceURL),
      }),
      [dbPool]
    );
  }
}
export default new CUDLConfig();
