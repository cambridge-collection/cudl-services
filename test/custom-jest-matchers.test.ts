import {BaseError} from '../src/errors';

describe('toThrowErrorTaggedWith', () => {
  function throwWithTags(...tags: string[]): never {
    throw new BaseError({tags});
  }

  async function promiseRejectedWithErrorTaggedWith(
    ...tags: string[]
  ): Promise<void> {
    throwWithTags(...tags);
  }

  test('passes when matching a single tag', () => {
    expect(() => throwWithTags('a')).toThrowErrorTaggedWith('a');
  });

  test('passes when matching a multiple tags', () => {
    expect(() => throwWithTags('a', 'b', 'c')).toThrowErrorTaggedWith([
      'a',
      'b',
      'c',
    ]);
  });

  test('passes when matching a subset of tags', () => {
    expect(() => throwWithTags('a', 'b', 'c', 'd')).toThrowErrorTaggedWith([
      'a',
      'b',
      'c',
    ]);
  });

  test('fails when callback is not a function', () => {
    expect(() =>
      expect(undefined).toThrowErrorTaggedWith('a')
    ).toThrowErrorMatchingSnapshot();
  });

  test('fails when callback does not throw', () => {
    expect(() =>
      expect(() => undefined).toThrowErrorTaggedWith('a')
    ).toThrowErrorMatchingSnapshot();
  });

  test('fails when thrown error is not tagged', () => {
    expect(() =>
      expect(() => {
        throw new Error();
      }).toThrowErrorTaggedWith('a')
    ).toThrowErrorMatchingSnapshot();
  });

  test('fails when thrown error does not have a single required tag', () => {
    expect(() =>
      expect(() => throwWithTags('b')).toThrowErrorTaggedWith('a')
    ).toThrowErrorMatchingSnapshot();
  });

  test('fails when thrown error is tagged with a subset of required tags', () => {
    expect(() =>
      expect(() => throwWithTags('a', 'b', 'c')).toThrowErrorTaggedWith([
        'a',
        'd',
      ])
    ).toThrowErrorMatchingSnapshot();
  });

  test('passes in .rejects mode when a promise rejects with a correctly-tagged error', async () => {
    await expect(
      promiseRejectedWithErrorTaggedWith('a')
    ).rejects.toThrowErrorTaggedWith('a');
  });

  test('fails in .rejects mode when a promise rejects with an incorrectly-tagged error', async () => {
    await expect(
      expect(
        promiseRejectedWithErrorTaggedWith('a')
      ).rejects.toThrowErrorTaggedWith('b')
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('fails in .rejects mode when a promise resolves', async () => {
    await expect(
      expect(Promise.resolve()).rejects.toThrowErrorTaggedWith('a')
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('fails in .resolves mode when a promise resolves', async () => {
    await expect(
      expect(Promise.resolve()).resolves.toThrowErrorTaggedWith('a')
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  test('fails in .resolves mode when a promise rejects', async () => {
    await expect(
      expect(Promise.resolve()).resolves.toThrowErrorTaggedWith('a')
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
