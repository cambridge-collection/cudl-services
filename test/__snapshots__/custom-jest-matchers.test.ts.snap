// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toThrowErrorTaggedWith fails in .rejects mode when a promise rejects with an incorrectly-tagged error 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mrejects[2m.[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m[\\"b\\"][39m to be in the error's tag set
Received: [31m[\\"a\\"][39m"
`;

exports[`toThrowErrorTaggedWith fails in .rejects mode when a promise resolves 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mrejects[2m.[22mtoThrowErrorTaggedWith[2m()[22m

Received promise resolved instead of rejected
Resolved to value: [31mundefined[39m"
`;

exports[`toThrowErrorTaggedWith fails in .resolves mode when a promise rejects 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mresolves[2m.[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Received value must be a function or Error but instead \\"[31mundefined[39m\\" was found"
`;

exports[`toThrowErrorTaggedWith fails in .resolves mode when a promise resolves 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mresolves[2m.[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Received value must be a function or Error but instead \\"[31mundefined[39m\\" was found"
`;

exports[`toThrowErrorTaggedWith fails when callback does not throw 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

expected an Error to be thrown but none was"
`;

exports[`toThrowErrorTaggedWith fails when callback is not a function 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Received value must be a function or Error but instead \\"[31mundefined[39m\\" was found"
`;

exports[`toThrowErrorTaggedWith fails when thrown error does not have a single required tag 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m[\\"a\\"][39m to be in the error's tag set
Received: [31m[\\"b\\"][39m"
`;

exports[`toThrowErrorTaggedWith fails when thrown error is not tagged 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m[\\"a\\"][39m to be in the error's tag set
Received: *error is not tagged*"
`;

exports[`toThrowErrorTaggedWith fails when thrown error is tagged with a subset of required tags 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoThrowErrorTaggedWith[2m([22m[32mexpected[39m[2m)[22m

Expected: [32m[\\"a\\", \\"d\\"][39m to be in the error's tag set
Received: [31m[\\"a\\", \\"b\\", \\"c\\"][39m"
`;
